File: frontend/src/settings/alertRules.js
// Task 42: "tell me when X happens here" -- the shape of a reader-defined
// alert rule, shared between the admin builder (AlertRulesSection.jsx), the
// toast (alertToastLogic.js) and mergeSettings (defaults.js). Mirrors
// backend/alert_rules.py's own Rule/geofence/condition shapes field for
// field -- the two are validated independently (this file guards what an
// operator's own browser writes into the shared config; alert_rules.py
// guards what the cache worker reads back out of it, since the file on disk
// is hand-editable), but they must agree on what a well-formed rule looks
// like or a rule built here would simply never fire there.
//
// RULE_LAYERS is a curated subset of entity_latest's own kind names
// (backend/storage.py), not every kind that table holds -- only the ones
// confirmed (by reading the actual record_snapshot() call in each source
// module, not from memory -- see this task's own report) to carry a stable
// per-row identity and a lat/lon a geofence can test. Ships and aircraft are
// the two seen most often, so they lead the list. Every key here is also a
// top-level key in /api/health's own response (health.ais, health.adsb, ...)
// -- describeRuleStatus below reads that entry to tell "watching, quiet"
// apart from "cannot be evaluated, its feed is down", so a layer added here
// without a matching /api/health entry would silently read as unhealthy
// for ever (fail closed, not fail open -- the honest direction to be wrong
// in, and exactly what backend/alert_rules.py's evaluate_rules does too:
// a layer with no fetched rows contributes nothing, never a false match).

import { STALE_AFTER_SECONDS } from "../utils/tempo";

export const RULE_LAYERS = [
  { key: "ais", label: "Ships (AIS)" },
  { key: "adsb", label: "Aircraft (ADS-B)" },
  { key: "gdelt_conflict", label: "Conflict events (GDELT)" },
  { key: "acled", label: "Conflict events (ACLED)" },
  { key: "jamming", label: "GPS/radio jamming (GPSJam)" },
  { key: "gfw_detections", label: "Satellite vessel detections (GFW)" },
  { key: "firms", label: "Fires / thermal anomalies (FIRMS)" },
];
const RULE_LAYER_KEYS = new Set(RULE_LAYERS.map((l) => l.key));

export const CONDITION_TYPES = [
  { key: "enter", label: "An entity is inside the area" },
  { key: "count_exceeds", label: "The count inside the area exceeds N" },
  { key: "score_above", label: "A field on an entity is above X" },
  { key: "squawk_equals", label: "An aircraft is squawking a given code" },
];

// "rect" is named for what it is -- a rectangle from the current viewport --
// rather than "draw a shape" or anything implying a hand-drawn polygon. See
// AlertRulesSection.jsx's own module note for why a real freehand-drawing
// tool was not built for this task, and its rect branch for the UI's own
// caveat alongside this label.
export const GEOFENCE_TYPES = [
  { key: "world", label: "Anywhere (no area)" },
  { key: "country", label: "A country" },
  { key: "region", label: "A named region" },
  { key: "water", label: "A sea, lake or river" },
  { key: "rect", label: "A rectangle (current map view)" },
];

function isPlainObject(value) {
  return !!value && typeof value === "object" && !Array.isArray(value);
}

function cleanStr(value, maxLen = 200) {
  if (typeof value !== "string") return null;
  const trimmed = value.trim();
  return trimmed ? trimmed.slice(0, maxLen) : null;
}

function cleanBbox(value) {
  if (!Array.isArray(value) || value.length !== 4) return null;
  const nums = value.map(Number);
  if (nums.some((n) => !Number.isFinite(n))) return null;
  const [south, west, north, east] = nums;
  if (south < -90 || south > 90 || north < -90 || north > 90) return null;
  if (west < -180 || west > 180 || east < -180 || east > 180) return null;
  if (south > north) return null;
  return nums;
}

/** A stored geofence -> a clean one, or null (meaning "the whole world",
 *  which is also what a rule with no geofence at all means). Region keys
 *  are checked against `regionKeys` -- the live REGIONS table from
 *  /api/regions -- rather than a list restated here, for the same "one
 *  definition of what a region is" reason regions.py's own REGIONS dict is
 *  shared between the region bar and the backend's query filter. */
export function sanitizeGeofence(raw, regionKeys) {
  if (raw == null) return null;
  if (!isPlainObject(raw)) return null;
  if (raw.type === "country") {
    const iso2 = cleanStr(raw.iso2, 2);
    return iso2 ? { type: "country", iso2: iso2.toUpperCase(), name: cleanStr(raw.name, 120) } : null;
  }
  if (raw.type === "region") {
    const key = cleanStr(raw.key, 60);
    if (!key || (regionKeys && !regionKeys.has(key))) return null;
    return { type: "region", key, name: cleanStr(raw.name, 120) };
  }
  if (raw.type === "water") {
    const id = cleanStr(raw.id, 120);
    const bbox = cleanBbox(raw.bbox);
    return id && bbox ? { type: "water", id, name: cleanStr(raw.name, 120), bbox } : null;
  }
  if (raw.type === "rect") {
    const bounds = cleanBbox(raw.bounds);
    return bounds ? { type: "rect", bounds } : null;
  }
  return null;
}

/** A stored condition -> a clean one, or null. */
export function sanitizeCondition(raw) {
  if (!isPlainObject(raw)) return null;
  if (raw.type === "enter") return { type: "enter" };
  if (raw.type === "count_exceeds") {
    const n = Math.round(Number(raw.n));
    return Number.isFinite(n) && n >= 0 ? { type: "count_exceeds", n } : null;
  }
  if (raw.type === "score_above") {
    const fieldName = cleanStr(raw.field, 60);
    const threshold = Number(raw.threshold);
    return fieldName && Number.isFinite(threshold) ? { type: "score_above", field: fieldName, threshold } : null;
  }
  if (raw.type === "squawk_equals") {
    const code = cleanStr(raw.code, 4);
    return code && /^\d+$/.test(code) ? { type: "squawk_equals", code } : null;
  }
  return null;
}

/**
 * A stored `alertRules` array -> the rules worth keeping in local state,
 * every malformed or unrecognised entry dropped rather than repaired --
 * same call sanitizeFilterPresets (defaults.js) makes for the same reason:
 * a rule missing a name or a condition is not a recognisable thing to
 * repair into, it is just not a rule. Mirrors backend/alert_rules.py's own
 * parse_rules() field for field (including the "squawk_equals only means
 * anything on the adsb layer" and "a geofence that fails to parse drops the
 * whole rule rather than silently widening it to the whole world" rules),
 * so a rule this function accepts is a rule the cache worker will actually
 * evaluate.
 */
export function sanitizeAlertRules(stored, regionKeys) {
  if (!Array.isArray(stored)) return [];
  const out = [];
  const seenIds = new Set();
  for (const entry of stored) {
    if (!isPlainObject(entry)) continue;
    const id = cleanStr(entry.id, 80);
    const name = cleanStr(entry.name, 120);
    const layer = RULE_LAYER_KEYS.has(entry.layer) ? entry.layer : null;
    const condition = sanitizeCondition(entry.condition);
    if (!id || seenIds.has(id) || !name || !layer || !condition) continue;
    if (condition.type === "squawk_equals" && layer !== "adsb") continue;
    let geofence = null;
    if (entry.geofence != null) {
      geofence = sanitizeGeofence(entry.geofence, regionKeys);
      if (!geofence) continue; // a broken geofence drops the whole rule -- see module note
    }
    seenIds.add(id);
    out.push({
      id,
      name,
      layer,
      enabled: entry.enabled !== false,
      geofence,
      condition,
      createdAt: Number.isFinite(entry.createdAt) ? entry.createdAt : Date.now(),
    });
  }
  return out;
}

/** A fresh, empty rule for the builder form to start from. Not itself valid
 *  (no condition has been chosen) until validateAlertRule below says so. */
export function blankAlertRule() {
  return {
    id: `rule-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`,
    name: "",
    layer: RULE_LAYERS[0].key,
    enabled: true,
    geofence: null,
    condition: { type: "enter" },
    createdAt: Date.now(),
  };
}

/** Whether a rule (as the builder form currently has it) is complete enough
 *  to save, and if not, the one thing to fix -- a single message rather
 *  than a field-by-field map, because the form has one save button and one
 *  place to show why it is disabled. */
export function validateAlertRule(rule) {
  if (!rule || !cleanStr(rule.name, 120)) return "Give the rule a name.";
  if (!RULE_LAYER_KEYS.has(rule.layer)) return "Choose a layer.";
  const condition = sanitizeCondition(rule.condition);
  if (!condition) {
    if (rule.condition?.type === "score_above") return "Give the field a name and a threshold.";
    if (rule.condition?.type === "squawk_equals") return "Choose a squawk code.";
    if (rule.condition?.type === "count_exceeds") return "Give the count a threshold.";
    return "Choose a condition.";
  }
  if (condition.type === "squawk_equals" && rule.layer !== "adsb") {
    return "Squawk codes only apply to the Aircraft (ADS-B) layer.";
  }
  if (rule.geofence != null && !sanitizeGeofence(rule.geofence, null)) return "That area is not set up yet.";
  return null;
}

/** Plain-language description of where a rule watches, for the rule list
 *  row and the toast -- the identical text backend/alert_rules.py's own
 *  _geofence_label builds server-side for the alert `detail`, so a reader
 *  never sees the admin panel and the toast describe the same rule two
 *  different ways. */
export function describeGeofence(geofence) {
  if (!geofence) return "anywhere";
  if (geofence.type === "country") return geofence.name || geofence.iso2;
  if (geofence.type === "region") return geofence.name || geofence.key;
  if (geofence.type === "water") return geofence.name || geofence.id;
  if (geofence.type === "rect") return "the marked area";
  return "the marked area";
}

/** Plain-language description of what a rule is watching for, for the same
 *  two call sites describeGeofence serves. */
export function describeCondition(condition) {
  if (!condition) return "";
  if (condition.type === "enter") return "is present";
  if (condition.type === "count_exceeds") return `count exceeds ${condition.n}`;
  if (condition.type === "score_above") return `${condition.field} above ${condition.threshold}`;
  if (condition.type === "squawk_equals") return `squawking ${condition.code}`;
  return "";
}

/**
 * The rule engine's own heartbeat (backend/app.py's `alert_rules` health
 * block) reduced to one label + severity class, for AlertRulesSection.jsx's
 * status line. Pulled out of that file (which is JSX and unreachable by
 * this project's `node --test` suite -- see frontend/tests/*.test.js) for
 * the same reason describeGeofence/describeCondition are: this composes a
 * user-visible sentence out of live data, and every task in this plan that
 * did that inline in JSX was sent back for it.
 *
 * Three of the four states this task's brief requires (never evaluated /
 * evaluated-and-quiet / currently firing / this rule's own layer unhealthy)
 * are told apart here or by the caller cross-referencing `health`'s other
 * entries; the fourth (firing) is a per-rule question the rule list row
 * answers on its own, not this section-wide summary.
 */
export function describeEngineStatus(health) {
  const row = health?.alert_rules;
  if (!row || row.last_success == null) {
    return { severity: "err", text: "never evaluated -- the cache worker has not run yet" };
  }
  if (row.last_error) return { severity: "err", text: row.last_error };
  return { severity: "ok", text: `evaluated ${row.seconds_since_success ?? 0}s ago` };
}

/**
 * Whether `info` (one /api/health entry, e.g. health.ais) is currently
 * landing data -- the identical rule SourceStatusSection.jsx's own dot
 * colouring has always used (`info.last_success && info.seconds_since_success
 * < STALE_AFTER_SECONDS`), read here rather than restated so a layer this
 * map already calls unhealthy on the Source status fold cannot read as
 * healthy on a rule that watches it. No second staleness threshold invented
 * for this -- see STALE_AFTER_SECONDS's own module note in utils/tempo.js.
 */
export function layerIsHealthy(info) {
  return !!(info && info.last_success && info.seconds_since_success < STALE_AFTER_SECONDS);
}

/**
 * A rule row's own status -- {severity, text} -- distinguishing the four
 * states a reader actually needs to tell apart, in priority order:
 *
 *   1. currently firing (this rule's own subject is in health.alerts)
 *   2. paused (the reader's own on/off switch, `rule.enabled === false`)
 *   3. its layer cannot be evaluated right now (health[rule.layer] is not
 *      landing data -- see layerIsHealthy above)
 *   4. watching, and quiet
 *
 * This is the fix for the review's own critical finding on this task: the
 * rule list previously decided a row's dot from firing + enabled alone, so
 * a rule whose layer's feed had broken -- entity_latest_with_ids returning
 * nothing not because the geofence is empty but because the data behind it
 * is gone -- rendered identically to a healthy rule that is correctly
 * quiet. Both said "green, watching". A reader has no way to tell "nothing
 * is happening" from "we stopped being able to see", which is exactly the
 * inversion this whole plan exists to prevent -- worse here than anywhere
 * else on this map, because the entire point of a rule is that its silence
 * is supposed to mean something.
 *
 * State 3 is checked before state 4 but after 1/2 deliberately: a rule that
 * is firing or paused has already answered the question a reader is asking
 * ("is this happening" / "did I turn this off"), and layering a second,
 * lower-priority caveat on top of either would be noise, not honesty --
 * backend/app.py's own `_alert_rules_health` docstring frames this the same
 * way, as something a reader cross-references, not a state that overrides
 * an already-informative one.
 */
export function describeRuleStatus(rule, health) {
  const firing = (health?.alerts || []).some((a) => a.subject === `rule:${rule.id}`);
  if (firing) return { severity: "err", text: "Currently firing", pulse: true };
  if (!rule.enabled) return { severity: "warn", text: "Paused" };
  if (!layerIsHealthy(health?.[rule.layer])) {
    const label = RULE_LAYERS.find((l) => l.key === rule.layer)?.label || rule.layer;
    return { severity: "warn", text: `Cannot be evaluated -- ${label}'s feed is down` };
  }
  return { severity: "ok", text: "Watching" };
}

/** The "Use <name>" capture button's own label for the country/water area
 *  pickers, or the prompt to click the map first when nothing is selected
 *  yet -- shared by both (the two buttons differ only in what "click ... on
 *  the map" names) so the wording is written once and stays in step. */
export function describeAreaCaptureLabel(kind, selectedName) {
  if (selectedName) return `Use ${selectedName}`;
  return kind === "water" ? "Click a sea, lake or river on the map first" : "Click a country on the map first";
}

/**
 * The confirmation note shown once an area has actually been captured
 * ("Set to Iran.", "Area captured from the current view."), or "" before
 * anything has been captured yet (AreaPicker's own callers only render this
 * when draft.geofence is already set, but the empty-string fallback keeps
 * this safe to call unconditionally too).
 *
 * Reuses describeGeofence for the name rather than reading
 * geofence.name/iso2/id directly -- one definition of "what to call this
 * geofence", not two that could say something different once a rule has
 * both a builder-time confirmation and a list-row/toast description.
 */
export function describeAreaCaptureConfirmation(geofence) {
  if (!geofence) return "";
  if (geofence.type === "rect") return "Area captured from the current view.";
  return `Set to ${describeGeofence(geofence)}.`;
}

File: backend/alert_rules.py
"""User-defined alert rules -- "tell me when X happens here" -- evaluated by
the cache worker's periodic loop and fired through the exact alerts table and
webhook backend/cacheworker already uses for source health (see
backend/cacheworker/__init__.py's own Alert dataclass and
backend/storage.py's record_alert/resolve_alerts/active_alerts). Nothing here
is a second notification system: reusing that de-duplication is what makes
"this rule has been true for six hours" one row and one webhook post rather
than one per poll -- the identical contract cacheworker already gives "Redis
is unreachable", extended from "is the infrastructure healthy" to "does this
reader-defined condition hold".

Where this runs, and why
-------------------------
A rule condition (entity enters a geofence, a layer's count in some area
exceeds N, an entity's numeric field is above X, an aircraft is squawking a
given code) needs live entity positions. entity_latest is the compact
current-state table built for exactly this; entity_history (the raw,
~11 GB, three-day movement log) must never be read on a request path, and
this module never touches it -- only entity_latest, through
storage.entity_latest_with_ids.

That leaves two candidate hosts for the evaluation itself: the backend (an
HTTP request path) or the cache worker (already a long-running, periodic,
off-request-path loop that evaluates conditions and fires them through this
exact alerts table). The backend is the wrong host for a periodic,
no-request-triggered loop -- FastAPI has no cron of its own, and building one
here would duplicate the cache worker's loop for no reason it can name. So
this is evaluated entirely inside the cache worker (see gather_and_evaluate
below, called from cacheworker/__main__.py's run_once), which already owns
"evaluate a condition, fire it through record_alert/resolve_alerts, notify a
webhook" once a minute.

The cache worker is a separate process from the API that serves the browser,
so firing an alert there cannot push a toast into an open tab directly --
there is no socket between them, and building one would be a second delivery
path for something the app already has one of. It does not need one: a fired
alert lands in the same `alerts` Postgres table /api/health already serves
(see app.py's health()), and the frontend's useHealth hook already polls
/api/health for every session, admin or not (frontend/src/hooks/useHealth.js,
15s in Admin Mode, 60s otherwise). A rule that starts firing shows up in that
poll's `alerts` array within one cycle, exactly like "Redis is unreachable"
does today, and AlertToast.jsx watches that array for rule-tagged entries it
has not shown yet. No new transport exists for this -- the toast rides the
delivery path the health alerts already have.

"Entity enters", without history
---------------------------------
Entering is normally a transition, and a transition needs a previous state to
compare against. This module keeps none. Instead, "enters" is treated as
level-triggered -- "this entity is currently inside the geofence" -- exactly
like "Redis is unreachable" is level-triggered: record_alert fires once when
a (rule, entity) pair starts being true and is silent for as long as it stays
true (its own de-duplication, not a second one built here), and
resolve_alerts closes it the moment a later pass no longer finds that entity
inside. A geofence exit followed by a later re-entry is therefore a fresh
alert -- the "resolves and re-fires" behaviour the brief requires -- for
free, because it is the identical mechanism cacheworker already uses for "the
cache stopped following this kind" recovering and breaking again.

There is deliberately no separate per-entity memo of "was this entity inside
last time". The active set handed to storage.resolve_alerts each pass *is*
that state: one row per currently-matching (rule, entity) pair, in Postgres,
cleared the moment it stops matching, and already correct across a worker
restart. An unbounded in-process dict tracking "every entity ever seen" is
exactly the defect this plan has produced twice elsewhere (see this task's
own brief); this design never needs one, because the alerts table already
*is* the bounded, durable record of "what is true right now".
"""

import logging
from dataclasses import dataclass

from backend import regions, storage
from backend.cacheworker import Alert

log = logging.getLogger("osint-globe.alert_rules")

# Every alert this module fires carries this prefix on `subject`, so a reader
# of the alerts table (or the frontend's AlertToast) can tell a user rule
# apart from a source-health condition without a second field -- the same way
# cacheworker's own SERVER constant marks conditions about the cache itself.
SUBJECT_PREFIX = "rule:"

# The condition types the rule builder offers, matching the brief's own list
# verbatim ("entity enters, count exceeds N, score above X, squawk equals
# Y"). Anything else in a stored rule is a hand-edited or stale file and is
# dropped by parse_rules rather than guessed at.
CONDITION_TYPES = {"enter", "count_exceeds", "score_above", "squawk_equals"}
GEOFENCE_TYPES = {"country", "region", "water", "rect"}


@dataclass(frozen=True)
class Rule:
    id: str
    name: str
    layer: str
    enabled: bool
    geofence: dict | None
    condition: dict


def _clean_str(value, max_len: int = 200) -> str | None:
    if not isinstance(value, str):
        return None
    value = value.strip()
    return value[:max_len] if value else None


def _clean_bbox(value):
    if not isinstance(value, (list, tuple)) or len(value) != 4:
        return None
    try:
        nums = [float(v) for v in value]
    except (TypeError, ValueError):
        return None
    south, west, north, east = nums
    if not (-90.0 <= south <= 90.0 and -90.0 <= north <= 90.0):
        return None
    if not (-180.0 <= west <= 180.0 and -180.0 <= east <= 180.0):
        return None
    if south > north:
        return None
    return nums


def _parse_geofence(raw) -> dict | None:
    """A stored geofence -> a clean dict, or None (meaning "the whole
    world") for anything that doesn't parse. A rule whose geofence has gone
    bad is not the same thing as a rule with no geofence -- see parse_rules'
    own note -- but this function's job is only the shape; parse_rules is
    what decides whether an unparseable geofence sinks the whole rule.
    """
    if raw is None:
        return None
    if not isinstance(raw, dict):
        return None
    kind = raw.get("type")
    if kind == "country":
        iso2 = _clean_str(raw.get("iso2"), 2)
        if not iso2:
            return None
        return {"type": "country", "iso2": iso2.upper(), "name": _clean_str(raw.get("name"), 120)}
    if kind == "region":
        key = _clean_str(raw.get("key"), 60)
        if not key or regions.bounds_for(key) is None:
            return None
        return {"type": "region", "key": key, "name": _clean_str(raw.get("name"), 120)}
    if kind == "water":
        wid = _clean_str(raw.get("id"), 120)
        bbox = _clean_bbox(raw.get("bbox"))
        if not wid or bbox is None:
            return None
        return {"type": "water", "id": wid, "name": _clean_str(raw.get("name"), 120), "bbox": bbox}
    if kind == "rect":
        bbox = _clean_bbox(raw.get("bounds"))
        if bbox is None:
            return None
        return {"type": "rect", "bounds": bbox}
    return None


def _parse_condition(raw) -> dict | None:
    if not isinstance(raw, dict):
        return None
    kind = raw.get("type")
    if kind == "enter":
        return {"type": "enter"}
    if kind == "count_exceeds":
        try:
            n = int(raw.get("n"))
        except (TypeError, ValueError):
            return None
        if n < 0:
            return None
        return {"type": "count_exceeds", "n": n}
    if kind == "score_above":
        field_name = _clean_str(raw.get("field"), 60)
        try:
            threshold = float(raw.get("threshold"))
        except (TypeError, ValueError):
            return None
        if not field_name:
            return None
        return {"type": "score_above", "field": field_name, "threshold": threshold}
    if kind == "squawk_equals":
        code = _clean_str(raw.get("code"), 4)
        if not code or not code.isdigit():
            return None
        return {"type": "squawk_equals", "code": code}
    return None


def parse_rules(raw_rules) -> list[Rule]:
    """A stored `alertRules` list (admin_config's own, deliberately-dumb
    JSON -- see backend/admin_config.py's module docstring) -> the Rules
    worth evaluating.

    Tolerant, the same way admin_config.load() is: this file is meant to be
    hand-editable, and one malformed entry must not take every other rule
    down with it. A rule that fails to parse is dropped and logged once per
    pass rather than raised -- the frontend's own rule builder only ever
    writes a well-formed shape (guarded selects and number inputs, never a
    free-text JSON field), so a drop here is expected to mean "stale/hand-
    edited file", not "the rule builder produced something wrong".
    """
    out = []
    if not isinstance(raw_rules, list):
        return out
    seen_ids = set()
    for entry in raw_rules:
        if not isinstance(entry, dict):
            continue
        rid = _clean_str(entry.get("id"), 80)
        name = _clean_str(entry.get("name"), 120)
        layer = _clean_str(entry.get("layer"), 40)
        condition = _parse_condition(entry.get("condition"))
        if not rid or rid in seen_ids or not name or not layer or condition is None:
            log.warning("Dropping an unparseable alert rule: %r", entry.get("id"))
            continue
        # squawk_equals only means anything against the ADS-B feed -- the
        # field it compares (payload["squawk"]) does not exist on any other
        # kind, so a rule of this type against another layer can never match
        # anything and would sit silently, indistinguishable from "correctly
        # evaluated, never met" (the honesty rule this plan keeps repeating).
        # Dropping it here instead is the same "found nothing must not look
        # like did not look" call applied one level up: a rule that cannot
        # possibly evaluate is not a rule, and the frontend never offers this
        # combination in the first place (see AlertRulesSection.jsx).
        if condition["type"] == "squawk_equals" and layer != "adsb":
            continue
        geofence_raw = entry.get("geofence")
        if geofence_raw is not None:
            geofence = _parse_geofence(geofence_raw)
            if geofence is None:
                # A geofence that fails to parse must drop the whole rule, not
                # silently widen it to "anywhere" -- a region key with a typo
                # or a country a hand-edit misspelled is a broken rule, and
                # firing it unbounded would be a *more* surprising failure
                # than not firing it at all.
                log.warning("Dropping alert rule %r: its geofence does not parse", rid)
                continue
        else:
            geofence = None
        seen_ids.add(rid)
        out.append(Rule(
            id=rid,
            name=name,
            layer=layer,
            enabled=entry.get("enabled") is not False,
            geofence=geofence,
            condition=condition,
        ))
    return out


def _entity_label(entity_id: str, payload: dict) -> str:
    """A human-readable name for one row: whatever this kind's own record
    calls itself, falling back to the one thing every row always has. Checked
    in this order because it is the same fallback chain
    squawkAlertsLogic.js's alertLabel already settled on for aircraft, and
    "name" is what ais.py's own ShipName field lands under (see
    backend/sources/ais.py's normalise_static)."""
    for field_name in ("name", "callsign", "registration"):
        value = payload.get(field_name)
        if value:
            return str(value)
    return entity_id


def _in_geofence(lat, lon, geofence: dict | None, country_index) -> bool:
    if geofence is None:
        return True
    if not isinstance(lat, (int, float)) or not isinstance(lon, (int, float)):
        return False
    kind = geofence["type"]
    if kind == "country":
        if country_index is None:
            return False
        found = country_index.country_at(lat, lon)
        return bool(found and found.get("iso2") == geofence["iso2"])
    if kind == "region":
        bounds = regions.bounds_for(geofence["key"])
        return bounds is not None and regions.bbox_contains_point(list(bounds), lat, lon)
    if kind == "water":
        return regions.bbox_contains_point(geofence["bbox"], lat, lon)
    if kind == "rect":
        return regions.bbox_contains_point(geofence["bounds"], lat, lon)
    return False


def _geofence_label(geofence: dict | None) -> str:
    if geofence is None:
        return "anywhere"
    kind = geofence["type"]
    if kind == "country":
        return geofence.get("name") or geofence["iso2"]
    if kind == "region":
        return geofence.get("name") or geofence["key"]
    if kind == "water":
        return geofence.get("name") or geofence["id"]
    if kind == "rect":
        return "the marked area"
    return "the marked area"


def _matching_rows(rule: Rule, rows: list[tuple[str, dict]], country_index) -> list[tuple[str, dict]]:
    """Every (entity_id, payload) in `rows` this rule's geofence admits, in
    the order storage handed them over. Shared by every per-entity condition
    below -- count_exceeds counts this list's length, enter/score_above/
    squawk_equals iterate it -- so the geofence test is written once."""
    out = []
    for entity_id, payload in rows:
        if _in_geofence(payload.get("lat"), payload.get("lon"), rule.geofence, country_index):
            out.append((entity_id, payload))
    return out


def evaluate_rules(rules: list[Rule], entities_by_layer: dict, *, country_index=None) -> list[Alert]:
    """Pure: every alert every enabled rule currently wants firing, computed
    from data already fetched. No I/O, no clock read, no persistence --
    record_alert/resolve_alerts (storage.py) are what turn this list into
    "newly firing" vs "still firing" vs "resolved"; this function only
    answers "is it true right now".

    `entities_by_layer` is {layer: [(entity_id, payload), ...]}, exactly
    storage.entity_latest_with_ids's own return shape, one call per distinct
    layer a rule references (see gather_and_evaluate). A layer no rule
    mentions is never fetched, and a layer that is missing from this dict
    (rather than merely empty) is treated as empty here -- both read as "no
    matches", which is correct: an unfetched layer contributes nothing to
    fire, and a genuinely empty one contributes nothing either.
    """
    alerts: list[Alert] = []
    for rule in rules:
        if not rule.enabled:
            continue
        rows = entities_by_layer.get(rule.layer) or []
        condition = rule.condition
        ctype = condition["type"]
        subject = f"{SUBJECT_PREFIX}{rule.id}"
        where = _geofence_label(rule.geofence)

        if ctype == "count_exceeds":
            matching = _matching_rows(rule, rows, country_index)
            count = len(matching)
            if count > condition["n"]:
                alerts.append(Alert(
                    subject, "count",
                    "warning",
                    f"\"{rule.name}\": {count} {rule.layer} in {where}, above the "
                    f"threshold of {condition['n']}.",
                ))
            continue

        if ctype == "enter":
            for entity_id, payload in _matching_rows(rule, rows, country_index):
                alerts.append(Alert(
                    subject, f"entity:{entity_id}",
                    "warning",
                    f"\"{rule.name}\": {_entity_label(entity_id, payload)} is in {where}.",
                ))
            continue

        if ctype == "score_above":
            field_name, threshold = condition["field"], condition["threshold"]
            for entity_id, payload in _matching_rows(rule, rows, country_index):
                value = payload.get(field_name)
                if not isinstance(value, (int, float)) or value <= threshold:
                    continue
                alerts.append(Alert(
                    subject, f"entity:{entity_id}",
                    "warning",
                    f"\"{rule.name}\": {_entity_label(entity_id, payload)} has {field_name} "
                    f"{value} in {where}, above {threshold}.",
                ))
            continue

        if ctype == "squawk_equals":
            code = condition["code"]
            for entity_id, payload in _matching_rows(rule, rows, country_index):
                if payload.get("squawk") != code:
                    continue
                alerts.append(Alert(
                    subject, f"entity:{entity_id}",
                    "warning",
                    f"\"{rule.name}\": {_entity_label(entity_id, payload)} is squawking "
                    f"{code} in {where}.",
                ))
            continue

    return alerts


async def gather_and_evaluate(rules_payload=None) -> list[Alert]:
    """The I/O half: read the stored rules and whatever entity_latest/country
    data they need, then hand it to the pure evaluate_rules above. Mirrors
    cacheworker/__main__.py's own split -- _probe does the I/O, evaluate() in
    cacheworker/__init__.py is pure -- one level higher, for the same reason:
    the part that decides what is wrong is the part that has to be right, and
    it is the part a test can reach without a Postgres.

    `rules_payload` is the parsed `alertRules` list from admin_config.load();
    the caller reads that (cacheworker/__main__.py) rather than this module
    doing it, so this stays a pure function of its arguments plus storage,
    the same testable shape the rest of this module keeps.
    """
    rules = parse_rules(rules_payload or [])
    enabled = [r for r in rules if r.enabled]
    if not enabled:
        return []

    layers = sorted({r.layer for r in enabled})
    entities_by_layer = {}
    for layer in layers:
        entities_by_layer[layer] = await storage.entity_latest_with_ids(layer)

    country_index = None
    if any(r.geofence and r.geofence["type"] == "country" for r in enabled):
        countries_geojson = await storage.reference("countries")
        country_index = regions.CountryIndex(countries_geojson or {})

    return evaluate_rules(enabled, entities_by_layer, country_index=country_index)
